### ISSUE TITLE
[no-empty-interface] Fix introduces compile error

### ISSUE THREAD

**Issue by pauldraper** [typescript-eslint/typescript-eslint#3648]
**Repro**

```json
{
  "rules": {
    "@typescript-eslint/no-empty-interface": "error"
  }
}
```

```ts
type Json = JsonArray | JsonNumber | JsonNull | JsonObject | JsonString;
interface JsonArray extends Array<Json> {};
type JsonNumber = number;
type JsonNull = null;
type JsonObject = { [key: string]: Json };
type JsonString = string;
```

**Expected Result**

I expected the fix to not introduce a compile erorr.

**Actual Result**

The fix produced

```ts
type Json = JsonArray | JsonNumber | JsonNull | JsonObject | JsonString;
type JsonArray = Array<Json>;
type JsonNumber = number;
type JsonNull = null;
type JsonObject = { [key: string]: Json };
type JsonString = string;
```

which is invalid

```
Type alias 'Json' circularly references itself.
Type alias 'JsonObject' circularly references itself.
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.10.1` |
| `@typescript-eslint/parser`        | `3.10.1` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.9.10` |
| `node`                             | `12.22.2` |

---

Really, I'm just going to agree with https://github.com/typescript-eslint/typescript-eslint/issues/1640#issuecomment-594450506

The assertion

> An interface declaring no members is equivalent to its supertype.

is just false, as demonstrated in #1640 and also this example, which itself has probably been written a hundred times over.

Interfaces and type aliases are similar, but not actually equivalent.

Maybe someone smarter than me sees a resolution other than disabling this rule (locally or the whole project), but that's the only resolution I see.

**Comment by bradzacher** [comment]
I get the same errors regardless of the form of `JSONArray`:

![image](https://user-images.githubusercontent.com/7462525/126689421-1c6fbb5b-dd7b-4584-b006-15cd5e1e0c31.png)
[repl](https://www.typescriptlang.org/play?#code/C4TwDgpgBAUgzgewHZQLy0UgggJxwQxCgB8NkA5AVwFsAjCHEspKgG1afmQHlaArCAGNgnTAGVgOAJZIA5gG4AUDOAMAZvkHQu2PISgQAHqqQATOFFwEQAHh0A+KAG8AvktCRmVOgzRQkNPQ47uDamGwc6AHsIZ46vALCfgBKQgg4pjZwkjKyADTM9rFhyBLScn7Z5QpAA)

![image](https://user-images.githubusercontent.com/7462525/126689489-991a460f-0b9a-4f45-bda2-e81b23fa54ce.png)
[repl](https://www.typescriptlang.org/play?#code/C4TwDgpgBAUgzgewHZQLy0UgggJxwQxCgB8NkA5AVwFsAjCHEspKgG1afmQHlaArCAGNgnTAGVgOAJZIA5gG4AUKEjNcBIunWEAPFyQA+JSuj6qdBmihIa9HMfCnMbDuhvsHq-bwHCrAJSEEHAATHThJGVkAGmYjZUdmCWk5KwiUhSA)

**Comment by pauldraper** [comment]
@bradzacher oops, I've update the example to be correct.

```ts
type JsonObject = { [key: string]: Json };
```

**Comment by rafaelss95** [comment]
Is it still a problem? Unless I'm missing something I couldn't reproduce it with [TS 4.4.3](https://www.typescriptlang.org/play?ts=4.4.3#code/C4TwDgpgBAUgzgewHZQLy0UgggJxwQxCgB8NkA5AVwFsAjCHEspKgG1afmQHlaArCAGNgnTAGVgOAJZIA5gG4AUAHplUGcAYAzfIOhdseQlAgAPTUgAmcKLgIgAPAYB8UAN4BfJaEjM7x9H9HF29wfUwqOgY0KCQaehxQ3wM2DnQ49iTwnn4hEXQ3KABtAGsIEAAuKDhJGVkAXSqDKC9FH2ykCWk5GJruhSA).

**Comment by bradzacher** [comment]
newer versions of TS have improved the recursive support for type declarations.
it looks like 4.4 is now advanced enough to handle this.
So yeah - I think we can close this.
